src/devices/SoftwareSpi/SoftwareSpi.cs

```diff
@@ -0,0 +1,199 @@
+            {
+                throw new ArgumentException(nameof(dataToWrite));
+            }
+
+            int bitLen = _settings.DataBitLength;
+            int lastBit = bitLen - 1;
+
+            using (ChipSelect)
+            {
+                for (int i = 0; i < dataToRead.Length; i++)
+                {
+                    byte readByte = 0;
+                    for (int j = 0; j < bitLen; j++)
+                    {
+                        using (BitTransfer)
+                        {
+                            int bit = _settings.DataFlow == DataFlow.MsbFirst ? lastBit - j : j;
+                            bool bitToWrite = ((dataToWrite[i] >> bit) & 1) == 1;
+                            if (ReadWriteBit(bitToWrite))
+                            {
+                                readByte |= (byte)(1 << bit);
+                            }
+                        }
+                    }
+
+                    dataToRead[i] = readByte;
+                }
+            }
+        }
+
+        private bool ReadWriteBit(bool bitToWrite)
+        {
+            _controller.Write(_mosi, bitToWrite);
+            return (bool)_controller.Read(_miso);
+        }
+
+        public override void Read(Span<byte> data)
+        {
+            Span<byte> dataToWrite = stackalloc byte[data.Length];
+            dataToWrite.Fill(0);
+            TransferFullDuplex(dataToWrite, data);
+        }
+
+        public override void Write(ReadOnlySpan<byte> data)
+        {
+            Span<byte> dataToRead = stackalloc byte[data.Length];
+            TransferFullDuplex(data, dataToRead);
+        }
+
+        public override void WriteByte(byte data)
+        {
+            Span<byte> outData = stackalloc byte[1];
+            outData[0] = data;
+            Write(outData);
+        }
+
+        public override byte ReadByte()
+        {
+            Span<byte> data = stackalloc byte[1];
+            Read(data);
+            return data[0];
+        }
+
+        public new void Dispose()
+        {
+            _controller?.Dispose();
+            _controller = null;
+            base.Dispose();
+        }
+
+        private ScopeData _bitTransfer;
+        private ScopeData _chipSelect;
+        private Scope BitTransfer => new Scope(_bitTransfer);
+        private Scope ChipSelect => new Scope(_chipSelect);
+
+        private class ScopeData
+        {
+            Action _enter, _exit;
+
+            public ScopeData(Action enter = null, Action exit = null)
+            {
+                _enter = enter;
+                _exit = exit;
+            }
+
+            public void Enter()
+            {
+                if (_enter != null)
+                {
+                    _enter();
+                }
+            }
+
+            public void Exit()
+            {
+                if (_exit != null)
+                {
+                    _exit();
+                }
+            }
+        }
+        private struct Scope : IDisposable
```

Comment:
Why is this not a class as well? Or why is ScopeData not a struct?

Reply:
Scope is constructed frequently so it's a struct so that the operation is cheap. ScopeData doesn't need to be fast as it's only constructed once per usage.